### DISCUSSION TITLE
Setting Custom `gasPrice`

### DISCUSSION THREAD

**Discussion by JasperTimm** [pcaversaccio/xdeployer#53] [Q&A] [answered]
Hi there,

I've recently been using xdeployer on Polygon. While for testnet (Mumbai) the auto gasPrice calculations seem to work fine, that's not the case on Polygon mainnet unfortunately. After multiple attempts of using auto calculated gasPrices failing I resorted to calling the Create2Deployer contract myself, with custom gasPrices, rather than using the plugin.

I couldn't find an easy way to set gasPrices whilst using the plugin, am I missing something obvious?

It sort of brings up another question, which I saw mentioned in another discussion - given this is a HardHat plugin, is there any reason not to have a more direct tie in to the HardHat config? Especially the network info which would include gas prices. Are there other networks which aren't supported as well in the HardHat config?

Thanks for creating the plugin!

- Jasper

**Comment by pcaversaccio** [comment, marked as answer]
@JasperTimm thanks for opening this discussion.

I've been actually aware of this problem for a while. The reason behind that issue is gas under-pricing by the Polygon's RPC providers (you would need to multiply by at least 2x the `maxFeePerGas` and `maxPriorityFeePerGas` to include a deployment successfully). I've tested all other chains and it's a Polygon-specific problem. 

Polygon has some serious issues with EIP-1559 it seems:
![image](https://user-images.githubusercontent.com/25297591/180843880-c6016f8c-f01d-4c48-a0ef-68f661a44f54.png)

Some points:
- I won't adjust the `xdeployer` code since this is a Polygon-specific problem and on all other 39 EVM chains it apparently works… - even L2 solutions like Optimism & Arbitrum have no issues so far (see e.g. [here](https://github.com/pcaversaccio/xdeployer/discussions/2#discussioncomment-1650713));
- Polygon RPCs will probably fix these issues soon and thereafter everything will work as intended. I encourage you to file an issue somewhere on Polygon's GitHub for that. @wschwab or @maxsam4 could you maybe forward this issue?;
- And the most important point: I don't want to enable customisation of the `gasPrice` etc. in `xdeployer`. First, this parameter would need to be adjusted for each chain (if EIP1559 is supported) and this leads to higher error rates (`gasLimit` is another topic because this is about the sequence of opcodes which is the same for all EVM chains). I could implement a direct patch in the code but that would mean that I'm calculating the fees and once the RPCs are adjusted the users are completely overpaying. Thereafter, I could ofc fix that again but all the people using the old version will still overpay meaning I would need to communicate hard that everyone needs to update. There would be a lot of problems as you see.

So the temporary solution I propose to you is either wait for Polygon to fix their RPC pricing issue or fork the code and patch it in the way you need it (UPDATE: see below how to patch the code). My goal is really to leave all the management to ethers, as [recommended by its author](https://github.com/ethers-io/ethers.js/issues/1610#issuecomment-915417595):
![image](https://user-images.githubusercontent.com/25297591/180845425-452326ee-f4cd-471c-8e77-21251ff56d9f.png)

Concerning your second point: I do see your points here, but my goal is to have a plugin that contains all the necessary information within its own config and is not affected by potential breaking changes from Hardhat's config structure. Or in other words, I think the design of the current plugin is more future-proof than making it dependent on core functionalities that can change over time. I want to assure that this plugin also works for hardhat versions 3+ without any major refactoring. The current structure also allows using specific RPCs for the `CREATE2` deployment without adjusting the `networks` config of Hardhat itself. Also, my plugin needs specific network names and can't rely on custom network names directly (that's also one of the reasons why you need to customise all the custom networks separately for [hardhat-etherscan](https://hardhat.org/hardhat-runner/plugins/nomiclabs-hardhat-etherscan#adding-support-for-other-networks)). In summary, I want to keep everything as much as possible isolated and offer the most flexibility regarding network configurations.

**Reply by JasperTimm** [reply]
Hey @pcaversaccio,

Thanks for the detailed response! That all makes sense.

I think the summary that essentially it's a Polygon specific problem and it will (hopefully) be fixed in the near future works.

Rather than go into an in-depth discussion regarding design decisions on whether it makes sense to add custom gas prices to xdeployer I think we'll probably just use our own fork for the time being.

Also - thanks for the code, that's essentially what I did when running it myself so far.

Cheers!

**Reply by 0xsharma** [reply]
Hey @JasperTimm @pcaversaccio ,

I hope you're doing well. I am from the Polygon Core-Protocol Team. Thank you for bringing this issue to light.
The EIP-1559 works fine on the consensus level, the gas prices are increasing/decreasing in expected manner. I also checked various RPC's and their gas estimation, it is in alignment with the nodes. Also, metamask is giving correct( verified from mainnet node) estimation of gas and the tx is also getting included instantly.

Can you please let us know if you're still facing this issue ? If yes, please provide with the steps/RPC which can reproduce the issue and we will look into it. 

Thank you 💜

**Reply by pcaversaccio** [reply]
@0xsharma the error is still here. I also want to make you aware of an open issue in the ethers.js repo [here](https://github.com/ethers-io/ethers.js/issues/2828). I shared with you over TG some debug information. In order for you to reproduce the error you can do the following:

1. Clone my repo here: https://github.com/pcaversaccio/batch-distributor
2. Replace in the [`hardhat.config.ts`](https://github.com/pcaversaccio/batch-distributor/blob/main/hardhat.config.ts#L331) the `xdeploy` configurations with the following:
```
  xdeploy: {
    contract: "BatchDistributor",
    constructorArgsPath: undefined,
    salt: "hello11!",
    signer: process.env.PRIVATE_KEY,
    networks: ["polygon"],
    rpcUrls: [
      "https://polygon-rpc.com",
    ],
    gasLimit: 2 * 10 ** 6,
  },
```
- Use a private key in the `.env` file with some Matic on it and run `yarn xdeploy`.
- Repeat with different other RPC providers (make sure you change the `salt` if one of the tx was successfully propagated).

I share now a typical error for Polygon deployments (and to re-stress, all other EVM chains work so it's not a plugin problem here):

```json
{"network":"polygon","contract":"BatchDistributor","address":"0x2f2502baB2Ea1f7d0394e26CF0a2B38911eE2212","deployed":false,"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":55,\"error\":{\"code\":-32000,\"message\":\"transaction underpriced\"}}","error":{"code":-32000},"requestBody":"{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x02f90b128189098459682f008459682f20832dc6c09413b0d85ccb8bf860b6b79af3029fca081ae9bef280b90aa466cfa0570000000000000000000000000000000000000000000000000000000000000000a723c03e36fb68bc771fe27da4adac1c429ee512a8382cd948150b6286a9913600000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000a1d6080604052610a0a806100136000396000f3fe6080604052600436106100295760003560e01c80633bd08a791461002e5780639d0918b514610050575b600080fd5b34801561003a57600080fd5b5061004e610049366004610799565b610063565b005b61004e61005e3660046107e9565b610172565b600061006f8280610826565b905090506000805b828110156100b8576100898480610826565b8281811061009957610099610895565b90506040020160200135826100ae91906108c4565b9150600101610077565b506100db73ffffffffffffffffffffffffffffffffffffffff851633308461031c565b60005b8281101561016b576101636100f38580610826565b8381811061010357610103610895565b6101199260206040909202019081019150610904565b6101238680610826565b8481811061013357610133610895565b905060400201602001358773ffffffffffffffffffffffffffffffffffffffff166103f89092919063ffffffff16565b6001016100de565b5050505050565b600061017e8280610826565b9050905060005b818110156102895760006101998480610826565b838181106101a9576101a9610895565b6101bf9260206040909202019081019150610904565b73ffffffffffffffffffffffffffffffffffffffff166101df8580610826565b848181106101ef576101ef610895565b9050604002016020013560405160006040518083038185875af1925050503d8060008114610239576040519150601f19603f3d011682016040523d82523d6000602084013e61023e565b606091505b5050905080610280576040517fdd74906f0000000000000000000000000000000000000000000000000000000081523060048201526024015b60405180910390fd5b50600101610185565b5047801561031757604051600090339083908381818185875af1925050503d80600081146102d3576040519150601f19603f3d011682016040523d82523d6000602084013e6102d8565b606091505b5050905080610315576040517fdd74906f000000000000000000000000000000000000000000000000000000008152306004820152602401610277565b505b505050565b60405173ffffffffffffffffffffffffffffffffffffffff808516602483015283166044820152606481018290526103159085907f23b872dd00000000000000000000000000000000000000000000000000000000906084015b604080517fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe08184030181529190526020810180517bffffffffffffffffffffffffffffffffffffffffffffffffffffffff167fffffffff000000000000000000000000000000000000000000000000000000009093169290921790915261044e565b60405173ffffffffffffffffffffffffffffffffffffffff83166024820152604481018290526103179084907fa9059cbb0000000000000000000000000000000000000000000000000000000090606401610376565b60006104b0826040518060400160405280602081526020017f5361666545524332303a206c6f772d6c6576656c2063616c6c206661696c65648152508573ffffffffffffffffffffffffffffffffffffffff1661055a9092919063ffffffff16565b80519091501561031757808060200190518101906104ce9190610921565b610317576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152602a60248201527f5361666545524332303a204552433230206f7065726174696f6e20646964206e60448201527f6f742073756363656564000000000000000000000000000000000000000000006064820152608401610277565b60606105698484600085610573565b90505b9392505050565b606082471015610605576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152602660248201527f416464726573733a20696e73756666696369656e742062616c616e636520666f60448201527f722063616c6c00000000000000000000000000000000000000000000000000006064820152608401610277565b73ffffffffffffffffffffffffffffffffffffffff85163b610683576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601d60248201527f416464726573733a2063616c6c20746f206e6f6e2d636f6e74726163740000006044820152606401610277565b6000808673ffffffffffffffffffffffffffffffffffffffff1685876040516106ac9190610967565b60006040518083038185875af1925050503d80600081146106e9576040519150601f19603f3d011682016040523d82523d6000602084013e6106ee565b606091505b50915091506106fe828286610709565b979650505050505050565b6060831561071857508161056c565b8251156107285782518084602001fd5b816040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102779190610983565b73ffffffffffffffffffffffffffffffffffffffff8116811461077e57600080fd5b50565b60006020828403121561079357600080fd5b50919050565b600080604083850312156107ac57600080fd5b82356107b78161075c565b9150602083013567ffffffffffffffff8111156107d357600080fd5b6107df85828601610781565b9150509250929050565b6000602082840312156107fb57600080fd5b813567ffffffffffffffff81111561081257600080fd5b61081e84828501610781565b949350505050565b60008083357fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe184360301811261085b57600080fd5b83018035915067ffffffffffffffff82111561087657600080fd5b6020019150600681901b360382131561088e57600080fd5b9250929050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052603260045260246000fd5b808201808211156108fe577f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b92915050565b60006020828403121561091657600080fd5b813561056c8161075c565b60006020828403121561093357600080fd5b8151801515811461056c57600080fd5b60005b8381101561095e578181015183820152602001610946565b50506000910152565b60008251610979818460208701610943565b9190910192915050565b60208152600082518060208401526109a2816040850160208701610943565b601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe016919091016040019291505056fea26469706673582212206614650dac716e200e5bf4b36a4a9f05ac46da675e8792d55dc202b9cbb25bb764736f6c63430008110033000000c080a0ba4080cf0e133bb9b7195cab422eee612558c7f58e1243b467eb6cb1520ccf42a014880d5d8a06c5a837a47b5355b60f730b5c0efc5e33998e4736869f5596dbb3\"],\"id\":55,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"https://polygon-rpc.com","transaction":{"type":2,"chainId":137,"nonce":9,"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f20"},"gasPrice":null,"gasLimit":{"type":"BigNumber","hex":"0x2dc6c0"},"to":"0x13b0D85CcB8bf860b6b79AF3029fCA081AE9beF2","value":{"type":"BigNumber","hex":"0x00"},"data":"0x66cfa0570000000000000000000000000000000000000000000000000000000000000000a723c03e36fb68bc771fe27da4adac1c429ee512a8382cd948150b6286a9913600000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000a1d6080604052610a0a806100136000396000f3fe6080604052600436106100295760003560e01c80633bd08a791461002e5780639d0918b514610050575b600080fd5b34801561003a57600080fd5b5061004e610049366004610799565b610063565b005b61004e61005e3660046107e9565b610172565b600061006f8280610826565b905090506000805b828110156100b8576100898480610826565b8281811061009957610099610895565b90506040020160200135826100ae91906108c4565b9150600101610077565b506100db73ffffffffffffffffffffffffffffffffffffffff851633308461031c565b60005b8281101561016b576101636100f38580610826565b8381811061010357610103610895565b6101199260206040909202019081019150610904565b6101238680610826565b8481811061013357610133610895565b905060400201602001358773ffffffffffffffffffffffffffffffffffffffff166103f89092919063ffffffff16565b6001016100de565b5050505050565b600061017e8280610826565b9050905060005b818110156102895760006101998480610826565b838181106101a9576101a9610895565b6101bf9260206040909202019081019150610904565b73ffffffffffffffffffffffffffffffffffffffff166101df8580610826565b848181106101ef576101ef610895565b9050604002016020013560405160006040518083038185875af1925050503d8060008114610239576040519150601f19603f3d011682016040523d82523d6000602084013e61023e565b606091505b5050905080610280576040517fdd74906f0000000000000000000000000000000000000000000000000000000081523060048201526024015b60405180910390fd5b50600101610185565b5047801561031757604051600090339083908381818185875af1925050503d80600081146102d3576040519150601f19603f3d011682016040523d82523d6000602084013e6102d8565b606091505b5050905080610315576040517fdd74906f000000000000000000000000000000000000000000000000000000008152306004820152602401610277565b505b505050565b60405173ffffffffffffffffffffffffffffffffffffffff808516602483015283166044820152606481018290526103159085907f23b872dd00000000000000000000000000000000000000000000000000000000906084015b604080517fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe08184030181529190526020810180517bffffffffffffffffffffffffffffffffffffffffffffffffffffffff167fffffffff000000000000000000000000000000000000000000000000000000009093169290921790915261044e565b60405173ffffffffffffffffffffffffffffffffffffffff83166024820152604481018290526103179084907fa9059cbb0000000000000000000000000000000000000000000000000000000090606401610376565b60006104b0826040518060400160405280602081526020017f5361666545524332303a206c6f772d6c6576656c2063616c6c206661696c65648152508573ffffffffffffffffffffffffffffffffffffffff1661055a9092919063ffffffff16565b80519091501561031757808060200190518101906104ce9190610921565b610317576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152602a60248201527f5361666545524332303a204552433230206f7065726174696f6e20646964206e60448201527f6f742073756363656564000000000000000000000000000000000000000000006064820152608401610277565b60606105698484600085610573565b90505b9392505050565b606082471015610605576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152602660248201527f416464726573733a20696e73756666696369656e742062616c616e636520666f60448201527f722063616c6c00000000000000000000000000000000000000000000000000006064820152608401610277565b73ffffffffffffffffffffffffffffffffffffffff85163b610683576040517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601d60248201527f416464726573733a2063616c6c20746f206e6f6e2d636f6e74726163740000006044820152606401610277565b6000808673ffffffffffffffffffffffffffffffffffffffff1685876040516106ac9190610967565b60006040518083038185875af1925050503d80600081146106e9576040519150601f19603f3d011682016040523d82523d6000602084013e6106ee565b606091505b50915091506106fe828286610709565b979650505050505050565b6060831561071857508161056c565b8251156107285782518084602001fd5b816040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102779190610983565b73ffffffffffffffffffffffffffffffffffffffff8116811461077e57600080fd5b50565b60006020828403121561079357600080fd5b50919050565b600080604083850312156107ac57600080fd5b82356107b78161075c565b9150602083013567ffffffffffffffff8111156107d357600080fd5b6107df85828601610781565b9150509250929050565b6000602082840312156107fb57600080fd5b813567ffffffffffffffff81111561081257600080fd5b61081e84828501610781565b949350505050565b60008083357fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe184360301811261085b57600080fd5b83018035915067ffffffffffffffff82111561087657600080fd5b6020019150600681901b360382131561088e57600080fd5b9250929050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052603260045260246000fd5b808201808211156108fe577f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b92915050565b60006020828403121561091657600080fd5b813561056c8161075c565b60006020828403121561093357600080fd5b8151801515811461056c57600080fd5b60005b8381101561095e578181015183820152602001610946565b50506000910152565b60008251610979818460208701610943565b9190910192915050565b60208152600082518060208401526109a2816040850160208701610943565b601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe016919091016040019291505056fea26469706673582212206614650dac716e200e5bf4b36a4a9f05ac46da675e8792d55dc202b9cbb25bb764736f6c63430008110033000000","accessList":[],"hash":"0x0b37f64b2f6183c90c21de20c46db369c1d7d67d58d5c993c6ab1ede4471973c","v":0,"r":"0xba4080cf0e133bb9b7195cab422eee612558c7f58e1243b467eb6cb1520ccf42","s":"0x14880d5d8a06c5a837a47b5355b60f730b5c0efc5e33998e4736869f5596dbb3","from":"0x9F3f11d72d96910df008Cfe3aBA40F361D2EED03","confirmations":0},"transactionHash":"0x0b37f64b2f6183c90c21de20c46db369c1d7d67d58d5c993c6ab1ede4471973c"}}
```

**Comment by pcaversaccio** [comment]
If you fork it, you could resolve the issue in the following manner (https://github.com/pcaversaccio/xdeployer/blob/main/src/index.ts#L144) - disclaimer: I'm hacking this together on my mobile lol:

```ts
createReceipt[i] = await create2Deployer[i].deploy(
  AMOUNT,
  hre.ethers.utils.id(hre.config.xdeploy.salt),
  initcode.data,
  {
    gasLimit: hre.config.xdeploy.gasLimit,
    maxPriorityFeePerGas: ethers.utils.parseUnits('YOUR_VALUE', 'gwei'),
    maxFeePerGas: ethers.utils.parseUnits('YOUR_VALUE', 'gwei'),
   }
);
```

**Reply by ahmadsghaier** [reply]
Great, thanks @pcaversaccio this helped me to quickly deploy on polygon mainnet with some hardcoded priority fees

**Reply by basememara** [reply]
These are the values that worked for me which were complete guesses but took awhile to get right, and may vary depending on network congestion:

```
gasLimit: hre.config.xdeploy.gasLimit, // 15_000_000 in hardhat.config
maxPriorityFeePerGas: ethers.utils.parseUnits('50', 'gwei'),
maxFeePerGas: ethers.utils.parseUnits('135', 'gwei')
```

I made these changes directly in `node_modules/xdeployer/dist/index.js` so I wouldn't have to fork and build.. quick and dirty hack. There are 2 places to make this change in the file but probably one of them only matters.

**Comment by pcaversaccio** [comment]
I would like to highlight that Polygon has an enforced minimum _30 gwei_ priority fee and `ethers.js` currently [hardcodes](https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-provider/src.ts/index.ts#L252) the priority fee, where applicable, at 1.5 gwei. That's why deployments on Polygon currently are failing/pending forever. There is an open issue that tracks this bug https://github.com/ethers-io/ethers.js/issues/3370. Once the issue is fixed, the `xdeployer` plugin will not have any issues anymore by default for Polygon deployments.

**Comment by basememara** [comment]
Apologies this is all very confusing even though I myself had to put in a workaround specifically for Polygon Mainnet's gas disaster. I tried modifying the source and even the dist files directly but just couldn't get it to work. It would've been great if someone had a PR as an example if everyone's saying it's so easy or "just do that". Polygon Mainnet's gas estimate is a train wreck, it's not a fun area, I get it, it's not your fault or my fault but makes consumers even more irritated makes things very broken for a major chain. Anyways thx for a great plugin, looks like it was fixed in ethers 6.7: https://github.com/ethers-io/ethers.js/issues/2828#issuecomment-1663273122

**Reply by pcaversaccio** [reply]
The latest `xdeployer` version uses the latest `ethers` version. Did you still encounter an issue, or did you use an older `xdeployer` version?

**Reply by basememara** [reply]
Ah ya you're right, just noticed I'm on v1. Tried jumping to v2 which forced ethers v6 and hardhat v-next which cascaded many things that got too far. I tried pushing thru but ethers v6 still has many issues with TypeScript and many other random areas. It wasn't easy and seems like I'm not the only one, but glad xdeployer is forward compatible, I just have to catch up. Thanks again and will revisit when I can get past the ethers v6 upgrade issues on my side.
